### PR TITLE
PLAT-827: Add testing docs

### PR DESCRIPTION
**Pull request by mplesec**
[JIRA ticket](https://reciprocitylabs.atlassian.net/browse/PLAT-827)

#### Description
PR adds documentation for testing.

#### ChangeLOG

- [ ] I have added all notable deployment/development environment (onprem!) changes to [CHANGELOG.md](https://github.com/reciprocity/zen-ui/blob/main/README.md)

#### Useful links
[Zen UI readme](https://github.com/reciprocity/zen-ui/blob/main/README.md)

[PR review guidelines](https://github.com/reciprocity/zengrc/blob/develop/doc/dev_environment_and_process/pull_request_reviews.md)

[Git commit guidelines](https://github.com/reciprocity/zengrc/blob/develop/doc/dev_environment_and_process/commit_guidelines.md)

### CODE REVIEW

File: docs/testing.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
+# Testing
+
+# Guidelines
+- In order to reduce flakiness, all tests must always be deterministic!
+  - Meaning when running a tests several times, the outcome must be the same.
```
Comment (chavlji) [inline docs/testing.md:5]: nit: 'a tests' - remove 'a' (plural)

File: docs/testing.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+# Types of tests
+- Unit
+  - These tests assert against the expected output of a given unit's functionality. Prefer to test pure functions as there is no need to test handle any side effects.
+  - These tests are located next to the source code.
+- Integration
+  - These tests check the correct functionality of an integration of units. Prefer to test only the minimum set of necessary components for a given integration test. The same goes for mocking/stubbing components.
```
Comment (chavlji) [inline docs/testing.md:11]: nit: 'an integration' - remove 'an'

File: docs/testing.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  - These tests are located next to the source code.
+- E2E and visual tests aka VRT (Visual Regression Testing)
+  - As this project provides isolated UI components it does not rely on any external APIs and there is no need to spin up or mock a database or API calls. However, E2E and visual tests should test the app from the perspective of how the end-user would use it, which in this case is interacting with the components through Storybook.
+  - Snapshot diffing is used for visual testing. In order to ensure there are no small differences in snapshots caused by (even slight) runtime environments (like OS fonts and scrollbars) and to guarantee the same environment on local machine and CI, visual testing is enabled only when running tests in headless mode in a Docker container.
+  - These tests are located in the projects global score, in this case all the tests, configs, and related files are located in the `./cypress` folder.
+
+# Technology
+We use [Jest](https://jestjs.io/) for unit and integration tests and [Cypress](https://www.cypress.io/) for E2E/.
```
Comment (chavlji) [inline docs/testing.md:19]: nit: 'E2E/' - remove '/'

**Review by chavlji — APPROVED** [review]
Very nice put!
Might be a good idea to move part of `testing.md` where you describe terminology to some global docs repo?

File: docs/testing.md (hunk 3 above)
Comment (mplesec) [reply docs/testing.md]: Hah, forgot about this. Fixed.

File: docs/testing.md (hunk 2 above)
Comment (mplesec) [reply docs/testing.md]: That's correct actually, changed the sentence a bit tho.

File: docs/testing.md (hunk 1 above)
Comment (mplesec) [reply docs/testing.md]: Tnx, I meant to write `a test` here.

**Comment by mplesec** [comment]
@chavlji regarding moving those descriptions out. Thought about it, but then realized that different project might have different definitions for these, as we have here with what an E2E test is in our case.

File: docs/testing.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+To run E2E and visual tests in Cypress:
+```bash
+# Runs Cypress in interactive mode, without visual testing
+yarn run test:e2e
+
+# Runs Cypress in headless mode, with visual testing
+yarn run test:e2e:headless
+# Runs Cypress in headless mode and updates snapshots for visual testing
+yarn run test:e2e:update
+```
```
Comment (Donaab) [inline docs/testing.md:46]: @mplesec I would add an explanation for one more script we have "test:e2e:interactive"

Comment (mplesec) [reply docs/testing.md:46]: It's the same as the `test:e2e` one. The `test:e2e` script is an alias for it.

**Pull request merged**